rclcpp_lifecycle/src/lifecycle_node_state_manager.cpp

```diff
@@ -0,0 +1,432 @@
+{
+  if (is_transitioning()) {
+    RCUTILS_LOG_ERROR(
+      "%s currently in transition, failing requested transition id %d.",
+      node_base_interface_->get_name(),
+      transition_id);
+    if (header) {
+      state_services_manager_hdl_->send_change_state_resp(header, false);
+    }
+    return RCL_RET_ERROR;
+  }
+
+  is_transitioning_.store(true);
+  header_ = header;
+  transition_id_ = transition_id;
+  rcl_ret_ = RCL_RET_OK;
+
+  unsigned int current_state_id;
+  {
+    std::lock_guard<std::recursive_mutex> lock(state_machine_mutex_);
+    constexpr bool publish_update = true;
+    if (rcl_lifecycle_state_machine_is_initialized(&state_machine_) != RCL_RET_OK) {
+      RCUTILS_LOG_ERROR(
+        "Unable to change state for state machine for %s: %s",
+        node_base_interface_->get_name(), rcl_get_error_string().str);
+      rcl_ret_error();
+      return rcl_ret_;
+    }
+
+    pre_transition_primary_state_ = State(state_machine_.current_state);
+
+    if (
+      rcl_lifecycle_trigger_transition_by_id(
+        &state_machine_, transition_id_, publish_update) != RCL_RET_OK)
+    {
+      RCUTILS_LOG_ERROR(
+        "Unable to start transition %u from current state %s: %s",
+        transition_id_, state_machine_.current_state->label,
+        rcl_get_error_string().str);
+      rcutils_reset_error();
+      rcl_ret_error();
+      return rcl_ret_;
+    }
+    current_state_id = get_current_state_id();
+    update_current_state_();
+  }
+
+  cb_return_code_ = execute_callback(
+    current_state_id,
+    pre_transition_primary_state_);
+  process_callback_resp(cb_return_code_);
+  return rcl_ret_;
+}
+
+void
+LifecycleNodeStateManager::process_user_callback_resp(
+  node_interfaces::LifecycleNodeInterface::CallbackReturn cb_return_code)
+{
+  constexpr bool publish_update = true;
+  unsigned int current_state_id;
+
+  cb_return_code_ = cb_return_code;
+  auto transition_label = get_label_for_return_code(cb_return_code);
+  {
+    std::lock_guard<std::recursive_mutex> lock(state_machine_mutex_);
+    if (
+      rcl_lifecycle_trigger_transition_by_label(
+        &state_machine_, transition_label, publish_update) != RCL_RET_OK)
+    {
+      RCUTILS_LOG_ERROR(
+        "Failed to finish transition %u: Current state is now: %s (%s)",
+        transition_id_,
+        state_machine_.current_state->label,
+        rcl_get_error_string().str);
+      rcutils_reset_error();
+      rcl_ret_error();
+      return;
+    }
+    current_state_id = get_current_state_id();
+
+    update_current_state_();
+  }
+
+  // error handling ?!
+  // TODO(karsten1987): iterate over possible ret value
+  if (cb_return_code == node_interfaces::LifecycleNodeInterface::CallbackReturn::ERROR) {
+    RCUTILS_LOG_WARN("Error occurred while calling transition function, calling on_error.");
+    auto error_cb_code = execute_callback(
+      current_state_id,
+      pre_transition_primary_state_);
+    process_callback_resp(error_cb_code);
```

Comment:
I see you have this function potentially called recursively. It looks like its the `update_current_state_` that should prevent the recursion but one needs to be careful.

Reply:
Good catch, probably worth having a check for circular transition and might align well with [this comment's second suggestion](https://github.com/tgroechel/rclcpp/pull/2#discussion_r1195136499). (the other possible way is to split `process_callback_resp` out but that was very clunky in prior implementation).

Reply:
https://github.com/tgroechel/rclcpp/commit/0a4a97db091cfe534b371d477711e0398ecd9378